File: client/src/api/study.ts
import client from "./client";

export type ReviewQuality = "again" | "hard" | "good" | "easy";

export interface StudyCardData {
  id: number;
  question: string;
  answer: string;
  video_id: number;
  video_title: string;
  folder_id: number | null;
  ease_factor: number;
  interval: number;
  repetitions: number;
  due_date: string;
  last_reviewed: string | null;
}

export interface ReviewResult {
  id: number;
  ease_factor: number;
  interval: number;
  repetitions: number;
  due_date: string;
  last_reviewed: string;
}

export interface StudyStatsData {
  total_cards: number;
  due_now: number;
  learned_today: number;
  new_cards: number;
  learning_cards: number;
  mature_cards: number;
}

export async function fetchDueCards(params?: {
  folder_id?: number;
  limit?: number;
}): Promise<StudyCardData[]> {
  const response = await client.get<StudyCardData[]>("/study/due", { params });
  return response.data;
}

export async function reviewCard(
  flashcardId: number,
  quality: ReviewQuality,
): Promise<ReviewResult> {
  const response = await client.post<ReviewResult>(`/study/review/${flashcardId}`, {
    quality,
  });
  return response.data;
}

export async function fetchStudyStats(): Promise<StudyStatsData> {
  const response = await client.get<StudyStatsData>("/study/stats");
  return response.data;
}

File: api/schemas/study.py
from datetime import datetime
from enum import Enum

from sqlmodel import SQLModel


class ReviewQuality(str, Enum):
    """User-facing rating when reviewing a card."""

    AGAIN = "again"  # quality 1 — forgot, restart
    HARD = "hard"    # quality 3 — remembered with difficulty
    GOOD = "good"    # quality 4 — normal recall
    EASY = "easy"    # quality 5 — effortless


class StudyCard(SQLModel):
    id: int
    question: str
    answer: str
    video_id: int
    video_title: str
    folder_id: int | None
    ease_factor: float
    interval: int
    repetitions: int
    due_date: datetime
    last_reviewed: datetime | None


class ReviewRequest(SQLModel):
    quality: ReviewQuality


class ReviewResponse(SQLModel):
    id: int
    ease_factor: float
    interval: int
    repetitions: int
    due_date: datetime
    last_reviewed: datetime


class StudyStats(SQLModel):
    total_cards: int
    due_now: int
    learned_today: int
    new_cards: int       # never reviewed
    learning_cards: int  # in progress (0 < repetitions < 3)
    mature_cards: int    # repetitions >= 3
